contracts/staking/PoSValidatorManager.sol

```diff
@@ -130,4 +140,155 @@ abstract contract PoSValidatorManager is IPoSValidatorManager, ValidatorManager
+
+        // Ensure the delegator is active
+        Delegator memory delegator = $._delegatorStakes[validationID][_msgSender()];
+        require(
+            delegator.status == ValidatorStatus.Active, "ValidatorManager: Delegator not active"
+        );
+        delegator.status = ValidatorStatus.PendingRemoved;
+        delegator.endedAt = uint64(block.timestamp);
+
+        $._delegatorStakes[validationID][_msgSender()] = delegator;
+
+        Validator memory validator = _getValidator(validationID);
+        require(validator.weight >= delegator.weight, "PoSValidatorManager: Invalid weight");
+
+        // Submit the message to the Warp precompile.
+        bytes memory setValidatorWeightPayload = ValidatorMessages
+            .packSetSubnetValidatorWeightMessage(
+            validationID, _getAndIncrementNonce(validationID), validator.weight - delegator.weight
+        );
+        $._pendingEndDelegatorMessages[validationID][_msgSender()] = setValidatorWeightPayload;
+        bytes32 messageID = WARP_MESSENGER.sendWarpMessage(setValidatorWeightPayload);
+
+        emit DelegatorRemovalInitialized(validationID, messageID, _msgSender(), block.timestamp);
+    }
+
+    function resendEndDelegation(bytes32 validationID, address delegator) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+        require(
+            $._pendingEndDelegatorMessages[validationID][delegator].length > 0
+                && $._delegatorStakes[validationID][delegator].status == ValidatorStatus.PendingRemoved,
+            "PoSValidatorManager: Delegator removal not pending"
+        );
+        WARP_MESSENGER.sendWarpMessage($._pendingEndDelegatorMessages[validationID][delegator]);
+    }
+
+    function completeEndDelegation(uint32 messageIndex, address delegator) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
+        // Unpack the Warp message
+        WarpMessage memory warpMessage = _getPChainWarpMessage(messageIndex);
+        (bytes32 validationID, uint64 nonce, uint64 weight) =
+            ValidatorMessages.unpackSubnetValidatorWeightUpdateMessage(warpMessage.payload);
+        require(
+            $._pendingEndDelegatorMessages[validationID][delegator].length > 0
+                && $._delegatorStakes[validationID][delegator].status == ValidatorStatus.PendingRemoved,
+            "PoSValidatorManager: Delegator removal not pending"
+        );
+        delete $._pendingEndDelegatorMessages[validationID][delegator];
+
+        // Update the validator weight and delegator status
+        Validator memory validator = _getValidator(validationID);
+        require(
+            validator.weight - $._delegatorStakes[validationID][delegator].weight == weight,
+            "PoSValidatorManager: Invalid weight"
+        );
+        require(validator.messageNonce >= nonce, "PoSValidatorManager: Invalid nonce");
```

Comment:
for nonce check we should make sure if we've seen a nonce greater than the received message, we don't double the effect, since the newer message would already account for older message

Reply:
Since the nonce is set by the validator manager, it should be safe to change this to `==`

Reply:
This will be a potentially worse UI if delegators are added at a fast rate, since only the most recent weight update message is eligible to be delivered.

Reply:
Actually, I think the implementation as is (with the removed weight check above) is sufficient. By checking `validator.messageNonce >= nonce`, we guarantee that we're only receiving Warp messages pertaining to `initialize` calls originating from the validator manager. The only state change in `completeInitializeDelegation` and `completeEndDelegation` use information provided in the Warp message, so replay protection is implicitly provided by the P-Chain.

Reply:
*Sigh* this is a bit more complicated actually. If we only check `validator.messageNonce >= nonce`, then if a higher nonce is delivered before a lower nonce, then delivering the lower nonce will revert the weight change from the higher nonce. This is what you originally called out, but I needed to think through it a bit myself 😅 . I'll play around with it some more.

Reply:
I've implemented nonce management that updates the contract's stored validator weight on `initializeRegisterDelegation`/`initializeEndDelegation`. This allows us to elide any special handling on delivery order for the corresponding `complete` functions. This is safe because we don't need the validator's total weight for anything when we `completeEndDelegation`, since the delegation rewards are entirely determined by the delegator's weight.